pageserver/src/tenant.rs

```diff
@@ -1779,7 +1775,15 @@ impl Tenant {
     /// - detach + ignore (freeze_and_flush == false)
     ///
     /// This will attempt to shutdown even if tenant is broken.
-    pub(crate) async fn shutdown(&self, freeze_and_flush: bool) -> Result<(), ShutdownError> {
+    ///
+    /// `shutdown_progress` is a [`completion::Barrier`] which later shutdown attempts can join to
+    /// await tenant's shutdown, returned as an error if caller was not the first caller to
+    /// initiate shutdown.
```

Comment:
```suggestion
    /// `shutdown_progress` is a [`completion::Barrier`] for the shutdown initiated by this call.
    /// If the tenant is already shutting down, we return a clone of the first shutdown call's `Barrier` in the `Err`.
    /// This not-first caller can use the returned barrier to coalesce with the ongoing shutdown.
    /// 
```

Reply:
I am not so sure about this as we have the type in the fn signature:
> `$param_name` is a $type for ...

What is the problem with "joining in" here and elsewhere? Also not sure if this version is really better.

Reply:
> IMO the current comment reads like the passed-in barrier is returned on Err (like when sending in to a channel fails).
> 
> But actually you get back the ongoing shutdown's Barrier.

Hmm ok had not picked this up at all, need to consider this.

Reply:
I think it does look like a good idea, but I didn't understand the problem from the diff. Please give rationale for your suggestions in future for faster processing.

Reply:
Fixed in [3a28c3f](https://github.com/neondatabase/neon/pull/4678/commits/3a28c3fbfb8be612b3e47a03a0ba176792e8c10c).